File: dfa/pass.py
import math
import copy
import numpy as np
import tensorflow.keras as keras
(x_train, y_train),(x_test, y_test) = keras.datasets.mnist.load_data()
import cv2
from mpi4py import MPI

##
##	Pass input trough the network
##

def test(a, weights, b, dim, perc):
    #INIT
    output_temp = []
    for i in dim:
        output_temp.append(np.zeros(i))
    len_dim = len(dim)-1
    per = 0

    #MPI INIT
    per_layer = int(size/len_dim) #how many processors per layer
    if rank != size:
        layer = int(rank/per_layer) #layer map
        part = rank - layer*per_layer #which part of layer
        layer_part_size = int(dim[layer+1]/per_layer)

    per = int(len(a)*perc/100)

    input = a[:per]
    output = b[:per]

    t = 0
    p = 0

    for ind in range(per):
        if rank != size:
            #INIT
            if layer == 0:
                net_in = input[ind]
            
            else:
                net_in = []
                for i in range(per_layer):
                    net_in = np.append(net_in, comm.recv(source = (layer-1)*per_layer + i))

            
            net_out = np.zeros(int(dim[layer+1]/per_layer))
            for i in range(int(dim[layer+1]/per_layer)):
                net_out[i] = weights[i][-1] #bias
                for j in range(dim[layer]):
                    net_out[i] += weights[i][j]*net_in[j] #weights * previous_layer_output

            if layer < len_dim-1:
                net_out[net_out < 0] = 0 #relu
            else:
                net_out = np.exp(net_out)/np.sum(np.exp(net_out)) #softmax

            if layer < len_dim-1:
                for i in range(per_layer):
                    comm.send(net_out, dest = (layer+1)*per_layer + i)
            
            else:
                comm.send(net_out, dest = size)

        if rank == size:
            res = []
            for i in range(per_layer):
                res = np.append(res, comm.recv(source = (len_dim-1)*per_layer + i))

        if rank == size:
            if ind % (per/10) == 0:
                print(f"Testing {p*10}%")
                p += 1
            if np.argmax(res) == output[ind][0]:
                t += 1
    if rank == size:
        print(f"Accuracy: {t}/{per}")

comm = MPI.COMM_WORLD
size = comm.Get_size() - 1
rank = comm.Get_rank()
#rank == size - final part of the network

input = x_test
k = 14/28
input = [cv2.resize(i, (0, 0), fx = k, fy = k) for i in input]
input = [i.flatten()/256 - 1/2 for i in input]

output = y_test
output = [i.flatten() for i in output]

dim = [14*14,10,10,10]
len_dim = len(dim)-1

#MPI INIT
per_layer = int(size/len_dim) #how many processors per layer
if rank != size:
    layer = int(rank/per_layer) #layer map
    part = rank - layer*per_layer #which part of layer
    layer_part_size = int(dim[layer+1]/per_layer)

if rank != size:
    weights = np.random.normal(0, 0.1, (int(dim[layer+1]/per_layer), (dim[layer]+1)))

if rank == size:
    weights = 0

test(input, weights, output, dim, 10)

File: dfa/pass.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <mpi.h>

#define size1 100
#define size2 100
#define in_len 14*14
#define len_dim 3

void print_arr(double *arr, int n){
    for(int i = 0; i < n; i++){
        printf("%lf ", arr[i]);
    }
    printf("\n");
}

int largest(double *arr, int n){
    int ind = 0;
    double max = arr[0];
    for(int i = 0; i < n; i++){
        if(max < arr[i]){
            ind = i;
            max = arr[i];
        }
    }
    return ind;
}

int main (int argc, char **argv){
    srand(time(NULL));
    /* MPI INIT */
    int size, rank;
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    size--; // rank == size -> final part of the network
    int dim[len_dim+1] = {in_len, 10, 10, 10};

    int per_layer = size/len_dim; // how many processors per layer
    int layer; // layer map
    int layer_dim;
    int prev_layer_dim = dim[len_dim]; // for rank == size
    int layer_part_size = 0;
    int prev_layer_part_size;
    if(rank != size){
        layer = rank/per_layer; // layer map
        layer_dim = dim[layer+1];
        prev_layer_dim = dim[layer];
        layer_part_size = layer_dim/per_layer;
        prev_layer_part_size = prev_layer_dim/per_layer;
    }
    
    /* LOAD DATA */
    FILE *fp_xtr, *fp_ytr, *fp_xts, *fp_yts;
    double x_train[size1][in_len], y_train[size1], x_test[size2][in_len], y_test[size2];
    fp_xtr = fopen("img/x_train.txt", "r");
    fp_ytr = fopen("img/y_train.txt", "r");
    fp_xts = fopen("img/x_test.txt", "r");
    fp_yts = fopen("img/y_test.txt", "r");
    for(int i = 0; i < size1; i++){
        for(int j = 0; j < in_len; j++){
            fscanf(fp_xtr, "%lf", &x_train[i][j]);
            fscanf(fp_xts, "%lf", &x_test[i][j]);
        }
        fscanf(fp_ytr, "%lf", &y_train[i]);
        fscanf(fp_yts, "%lf", &y_test[i]);
    }
    fclose(fp_xtr);
    fclose(fp_ytr);
    fclose(fp_xts);
    fclose(fp_yts);

    /* INIT */
    //double weights[layer_part_size][prev_layer_dim+1]; //dyn alloc
    //double *weights = (double*) malloc(layer_part_size * (prev_layer_dim+1) * sizeof(double));
    double **weights = (double**) malloc(layer_part_size * sizeof(double *));
    if(rank != size){
        for(int i = 0; i < layer_part_size; i++){
            weights[i] = (double*) malloc((prev_layer_dim+1) * sizeof(weights));
        }

        for(int i = 0; i < layer_part_size; i++){
            for(int j = 0; j < prev_layer_dim+1; j++){
                weights[i][j] = (double(rand())/RAND_MAX - 0.5) / 4; // random
            }
        }
    }
    //double net_in[prev_layer_dim]; // dyn_alloc
    double *net_in = (double*) malloc(prev_layer_dim * sizeof(double));
    
    /* TRAIN */
    int perc = 100;
    int per = size1*perc/100;
    int t = 0, p = 0;
    int tag = 0; // TODO image count?

    for(int i = 0; i < per; i++){
        if(rank != size){
            // INIT
            if(layer == 0){
                for(int j = 0; j < in_len; j++){ // in_len = dim[0]
                    net_in[j] = x_train[i][j];
                }
            }
            else{
                for(int j = 0; j < per_layer; j++){
                    MPI_Recv(&net_in[j*prev_layer_part_size], prev_layer_part_size, MPI_DOUBLE, per_layer*(layer-1)+j, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
                }
            }

            double *net_out = (double*) malloc(layer_part_size * sizeof(double));

            for(int j = 0; j < layer_part_size; j++){
                net_out[j] = weights[j][layer_dim]; // bias
                for(int k = 0; k < prev_layer_dim; k++){
                    net_out[j] += weights[j][k] * net_in[k]; // weights * previous_layer_output
                }
            }

            if(layer < len_dim - 1){
                // ACTIVATION
                for(int j = 0; j < layer_part_size; j++){
                    if(net_out[j] < 0){
                        net_out[j] = 0; //relu
                    }
                }
                // FEED FORWARD 
                for(int j = 0; j < per_layer; j++){
                    MPI_Send(net_out, layer_part_size, MPI_DOUBLE, per_layer*(layer+1)+j, tag, MPI_COMM_WORLD);
                }
            }
            else{
                MPI_Send(net_out, layer_part_size, MPI_DOUBLE, size, tag, MPI_COMM_WORLD);
            }
            //TODO SOFTMAX
        }
        else{
            for(int j = 0; j < per_layer; j++){
                MPI_Recv(&net_in[j*dim[len_dim]/per_layer], dim[len_dim]/per_layer, MPI_DOUBLE, per_layer*(len_dim-1)+j, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
            }
            if(i % (per/10) == 0){
                printf("Testing %d..\n", p*10);
                p++;
            }
            if(largest(net_in, dim[len_dim]) == y_train[i]){
                t++;
            }
        }
    }
    if(rank == size){
        printf("Accuracy: %d/%d\n", t, per);
    }

    MPI_Finalize();
}

